Allow range filter on subfields of structs with coercion

mbasmanova opened the pull request:
Fixes `UnsupportedOperationException` for queries with range filters on struct subfields when applied to tables with updated schema.

```
Caused by: java.lang.UnsupportedOperationException: Range filers on array types are not supported
	at com.facebook.presto.hive.HiveCoercer$StructCoercer.toCoercingFilter(HiveCoercer.java:479)
	at com.facebook.presto.hive.orc.OrcSelectivePageSourceFactory.toTupleDomainFilters(OrcSelectivePageSourceFactory.java:545)
```

```
== NO RELEASE NOTE ==
```

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercer.java:
```diff
@@ -128,8 +132,9 @@ public Block apply(Block block)
-        public TupleDomainFilter toCoercingFilter(TupleDomainFilter filter)
+        public TupleDomainFilter toCoercingFilter(TupleDomainFilter filter, Subfield subfield)
         {
+            checkArgument(subfield.getPath().isEmpty(), "Subfields on primitive types are not allowed");
```
Do we wanna have a helper function called `isPrimitive()` in `Subfield` given it will be used multiple times?

mbasmanova replied:
@highker Something like this would be useful, but I can't think of a good name just yet. `isPrimitive` doesn't work because `a IS NULL` might represent column `a` of any type, primitive or not.

highker commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercer.java:
```diff
@@ -470,13 +480,23 @@ else if (i < rowBlock.getFieldCount()) {
+                        return coercers[i].toCoercingFilter(filter, subfield.tail(fieldName));
+                    }
+                }
+                verify(false, "Unexpected struct field: " + fieldName);
```
throw an exception instead

mbasmanova commented:
@highker James, thank you for review. 

- I changed `verify` call to `throw new IllegalArgumentException("Struct field not found: " + fieldName);`.
- I also noticed that range filters on newly added struct fields didn't work and fixed that.

wenleix commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveCoercer.java:
```diff
@@ -470,13 +479,23 @@ else if (i < rowBlock.getFieldCount()) {
+                throw new IllegalArgumentException("Struct field not found: " + fieldName);
+            }
+
+            throw new UnsupportedOperationException("Range filers on struct types are not supported");
```
curious: what about map?

mbasmanova commented:
@wenleix Wenlei, thank you for review.

> my understanding is it allows range filters on subfield of arrays right? And struct and map is still yet to be supported? Is my understanding correct ?

As of today, we support range filters on subfields of arrays and structs, but not maps. By default, we only enable range filters for subfields of structs though. E.g. a.b.c.d < 10 is supported, but a[3] = 5 is not.

mbasmanova commented:
@wenleix Just to clarify that range filters on subfields are separate from subfield pruning. We support subfield pruning for all types, e.g. arrays, maps and structs.
